fluidanimate/src-rl/cellpool.cpp

```diff
@@ -0,0 +1,317 @@
+    addr = (void *)&(cell->a[i]);
+    cell->a[i] = coord_const + a_vec_const + (uint64_t)addr;
+    addr = (void *)&(cell->density[i]);
+    cell->density[i] = density_vec_const + (uint64_t)addr;
+  }
+}
+
+//check whether a cell filled with write_cell has still the same data
+//returns true if the data has not changed
+bool check_cell(struct Cell *cell) {
+  Vec3 dummy;
+  bool rv = true;
+  int i;
+
+  //check every element of cell
+  for(i=0; i<PARTICLES_PER_CELL; i++) {
+    void *addr;
+
+    //values depend only on memory location and constants
+    addr = (void *)&(cell->p[i]);
+    dummy = coord_const + p_vec_const + (uint64_t)addr;
+    rv = rv && (dummy == cell->p[i]);
+    addr = (void *)&(cell->hv[i]);
+    dummy = coord_const + hv_vec_const + (uint64_t)addr;
+    rv = rv && (dummy == cell->hv[i]);
+    addr = (void *)&(cell->v[i]);
+    dummy = coord_const + v_vec_const + (uint64_t)addr;
+    rv = rv && (dummy == cell->v[i]);
+    addr = (void *)&(cell->a[i]);
+    dummy = coord_const + a_vec_const + (uint64_t)addr;
+    rv = rv && (dummy == cell->a[i]);
+    addr = (void *)&(cell->density[i]);
+    float fdummy = density_vec_const + (uint64_t)addr;
+    rv = rv && (fdummy == cell->density[i]);
+  }
+
+  return rv;
+}
+
+
+//A simple self-test program that stresses the cell pool a little
+//The basic strategy is to allocate lots of cells and write pre-determined values to
+//the cell arrays. Then we read the values to make sure they haven't been overwritten.
+int main() {
+  int nCells = 2 * 1000 * 1000; //test with 2 million cells
+  const int size_array = 389; //number of statically allocated cells (a prime number)
+  struct Cell cells[size_array]; //array of dummy cells, serves as entry points for lists
+  cellpool *pool;
+  int i;
+
+  printf("Initializing...\n");fflush(NULL);
+
+   //init with low number of particles to cause lots of work
+  cellpool_init(&pool, 1);
+
+  //initialize statically allocated cells
+  for(i=0; i<size_array; i++) {
+    write_cell(&(cells[i]));
+    cells[i].next = NULL;
+  }
+
+  printf("Testing (1st pass): ");fflush(NULL);
+
+  //allocate all cells and initialize them with predetermined values
+  for(i=0; i<nCells; i++) {
+    struct Cell *ptr, *temp;
+
+    //get a new cell and append it to lists in round-robin way
+    temp = cellpool_getcell(pool);
+    write_cell(temp);
+    ptr = &(cells[i % size_array]);
+    while(ptr->next != NULL) {
+      ptr = ptr->next;
+    }
+    ptr->next = temp;
+
+    //print a progress message every 1/10th of the work
+    if((i+1) % (nCells/10) == 0) {
+      printf("...%i%%", (i+1) / (nCells/100));fflush(NULL);
+    }
+  }
+  printf("\n");fflush(NULL);
+
+  //now make a 2nd pass and check that all values are still correct
+  printf("Testing (2nd pass): ");fflush(NULL);
+  int count = 0;
+  for(i=0; i<size_array; i++) {
+    struct Cell *ptr;
+
+    ptr = &(cells[i]);
+    do {
+      if(!check_cell(ptr)) {
```

Comment:
odd that things like the check_cell wouldn't be asserts if they result in an exit. Might change, I'll leave that one up to you.